pkg/cloud/gcp/gcs_kms.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"hash/crc32"
+	"net/url"
+	"strings"
+
+	kms "cloud.google.com/go/kms/apiv1"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/errors"
+	"google.golang.org/api/option"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+const gcsScheme = "gcs"
+
+type gcsKMS struct {
+	kms                 *kms.KeyManagementClient
+	customerMasterKeyID string
+}
+
+var _ cloud.KMS = &gcsKMS{}
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeGCSKMS, gcsScheme)
+}
+
+type kmsURIParams struct {
+	credentials string
+	auth        string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	params := kmsURIParams{
+		credentials: kmsURI.Query().Get(CredentialsParam),
+		auth:        kmsURI.Query().Get(cloud.AuthParam),
+	}
+
+	return params
+}
+
+// MakeGCSKMS is the factory method which returns a configured, ready-to-use
+// GCS KMS object.
+func MakeGCSKMS(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
+	if env.KMSConfig().DisableOutbound {
+		return nil, errors.New("external IO must be enabled to use GCS KMS")
+	}
+	kmsURI, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the URI parameters required to setup the GCS KMS session.
+	kmsURIParams := resolveKMSURIParams(*kmsURI)
+
+	// Client options to authenticate and start a GCS KMS session.
+	// Currently only accepting json of service account
+	var credentialsOpt []option.ClientOption
+
+	switch kmsURIParams.auth {
+	case "", cloud.AuthParamSpecified:
+		if kmsURIParams.credentials == "" {
+			return nil, errors.Errorf(
+				"%s is set to '%s', but %s is not set",
+				cloud.AuthParam,
+				cloud.AuthParamSpecified,
+				CredentialsParam,
+			)
+		}
+
+		// Credentials are passed in base64 encoded, so decode the credentials first
```

Comment:
[super nit]: here and if you can audit all comments and end them with a period, thanks!